app/controllers/task_contributions_controller.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+class TaskContributionsController < ApplicationController
+  include TaskParameters
+
+  before_action :load_and_authorize_base_task
+  before_action :load_and_authorize_task_contribution, except: %i[index create new]
+
+  def approve_changes
+    if @task.apply_contribution(@task_contribution)
+      TaskContributionMailer.approval_info(@task_contribution).deliver_later
+      redirect_to @task, notice: t('.success')
+    else
+      redirect_to [@task, @task_contribution], alert: t('.error')
+    end
+  end
+
+  def discard_changes
+    duplicate = @task_contribution.suggestion.duplicate(set_parent_identifiers: false)
+    if duplicate.save && @task_contribution.close
+      TaskContributionMailer.rejection_info(@task_contribution).deliver_later
+      self_closed = @task_contribution.user == current_user
+      redirect_to (self_closed ? duplicate : @task), notice: t('.success')
+    else
+      redirect_to [@task, @task_contribution], alert: t('.error')
+    end
+  end
+
+  def index
+    authorize @task, :edit?
+    @task_contributions = @task.contributions(all_states: true)
+  end
+
+  def show
+    @task = @task_contribution.suggestion
+
+    @files = @task.files
+    @tests = @task.tests
+    @model_solutions = @task.model_solutions
+
+    @user_rating = @task.ratings&.find_by(user: current_user)&.rating
+    render 'tasks/show'
+  end
+
+  def new
+    raise Pundit::NotAuthorizedError if current_user.nil?
+
+    @task_contribution = TaskContribution.new_for(@task, current_user)
+    authorize @task_contribution
+
+    @task = @task_contribution.suggestion
+    render 'task_contributions/new'
```

Comment:
This should be the default `render` action, so that you can probably omit this line.